f3.go

```diff
@@ -145,412 +78,197 @@ func New(ctx context.Context, manifest manifest.ManifestProvider, ds datastore.D
-			}
-			errCh <- err
-		}
-	}()
-
-	m.handleIncomingMessages(ctx, messageQueue)
+	return cs.Latest(), nil
 }
 
-func (m *F3) boostrap(ctx context.Context) error {
-	head, err := m.ec.GetHead(ctx)
-	if err != nil {
-		return xerrors.Errorf("failed to get the head: %w", err)
-	}
-
-	if head.Epoch() < m.Manifest().BootstrapEpoch {
-		// wait for bootstrap epoch
-		for {
-			head, err := m.ec.GetHead(ctx)
-			if err != nil {
-				return xerrors.Errorf("getting head: %w", err)
-			}
-			if head.Epoch() >= m.Manifest().BootstrapEpoch {
-				break
-			}
-
-			m.log.Infof("wating for bootstrap epoch (%d): currently at epoch %d", m.Manifest().BootstrapEpoch, head.Epoch())
-			aim := time.Until(head.Timestamp().Add(m.Manifest().ECPeriod))
-			// correct for null epochs
-			for aim < 0 {
-				aim += m.Manifest().ECPeriod
-			}
-
-			select {
-			case <-time.After(aim):
-			case <-ctx.Done():
-				return ctx.Err()
-			}
-		}
-	}
+func (m *F3) GetCert(ctx context.Context, instance uint64) (*certs.FinalityCertificate, error) {
+	m.mu.Lock()
+	cs := m.cs
+	m.mu.Unlock()
 
-	ts, err := m.ec.GetTipsetByEpoch(ctx, m.Manifest().BootstrapEpoch-m.Manifest().ECFinality)
-	if err != nil {
-		return xerrors.Errorf("getting initial power tipset: %w", err)
+	if cs == nil {
+		return nil, xerrors.Errorf("F3 is not running")
 	}
+	return cs.Get(ctx, instance)
+}
 
-	initialPowerTable, err := m.ec.GetPowerTable(ctx, ts.Key())
-	if err != nil {
-		return xerrors.Errorf("getting initial power table: %w", err)
+// Returns the time at which the F3 instance specified by the passed manifest should be started, or
+// 0 if the passed manifest is nil.
+func (m *F3) computeBootstrapDelay(manifest *manifest.Manifest) (delay time.Duration, err error) {
+	if manifest == nil {
+		return 0, nil
 	}
 
-	cs, err := certstore.CreateStore(ctx, m.ds, m.Manifest().InitialInstance, initialPowerTable)
+	ts, err := m.ec.GetHead(m.runningCtx)
 	if err != nil {
-		return xerrors.Errorf("creating certstore: %w", err)
-	}
-	m.setCertStore(cs)
-	return nil
-}
-
-func (m *F3) GetLatestCert(ctx context.Context) (*certs.FinalityCertificate, error) {
-	m.mLk.Lock()
-	defer m.mLk.Unlock()
-	if m.certStore == nil {
-		return nil, xerrors.Errorf("F3 is not running")
-	}
-	return m.certStore.Latest(), nil
-}
-func (m *F3) GetCert(ctx context.Context, instance uint64) (*certs.FinalityCertificate, error) {
-	m.mLk.Lock()
-	defer m.mLk.Unlock()
-	if m.certStore == nil {
-		return nil, xerrors.Errorf("F3 is not running")
-	}
-	return m.certStore.Get(ctx, instance)
+		err := xerrors.Errorf("failed to get the EC chain head: %w", err)
+		return 0, err
+	}
+
+	// We bootstrap now if we're at the correct time, whether or
+	// not we've hit the correct epoch. The important thing is
+	// that:
+	//
+	// 1. All nodes will use BootstrapEpoch - Finality to pick the base.
+	// 2. All nodes will bootstrap at the same time.
+	currentEpoch := ts.Epoch()
+	if currentEpoch < manifest.BootstrapEpoch {
```

Comment:
Maybe invert the if: return 0 `if currentEpoch >= manifest.BoostrapEpoch` then the more complex chunk is outside of if